{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Filtering & Plots\n",
    "\n",
    "This notebook is for streamlining the filtering process that occurs after ipyrad. This way, I can rerun the notebook with different filenames instead of do it from scratch each time.\n",
    "\n",
    "\n",
    "### Naming Cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "assembly_name = \"parcal_003\"\n",
    "filter_biall = \"TRUE\" # either TRUE or FALSE\n",
    "filter_MAF = \"TRUE\" # either TRUE or FALSE\n",
    "dir_outfiles = \"/mnt/hgfs/SHARED_FOLDER/Learn_iPyrad/PARCAL_RUN1/parcal_003_outfiles/\"\n",
    "date = \"20170903\"\n",
    "plot_subtitle = \"\" # leave blank if you don't want to add a subtitle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Working Cell:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals after adding fpop = 80.0\n",
      "Number of loci = 6362\n"
     ]
    }
   ],
   "source": [
    "# import all necessary modules\n",
    "import os\n",
    "import subprocess as sp\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# name all intermediate files\n",
    "ipy_VCF = assembly_name + \".vcf\" # e.g., CG_l1_25_c86_H.vcf\n",
    "filt_VCF_out = assembly_name # e.g., CG_l1_25_c86_H\n",
    "if filter_biall == \"TRUE\": \n",
    "    filt_VCF_out += \"_biall\" # e.g., CG_l1_25_c86_H_biall\n",
    "if filter_MAF == \"TRUE\":\n",
    "    filt_VCF_out += \"_maf\" # e.g., CG_l1_25_c86_H_biall_maf\n",
    "else:\n",
    "    filt_VCF_out += \"_NOmaf\" # e.g., CG_l1_25_c86_H_biall_NOmaf\n",
    "filt_VCF_out_recode = filt_VCF_out + \".recode.vcf\" # e.g.,CG_l1_25_c86_H_biall_maf.recode.vcf\n",
    "one_VCF = filt_VCF_out.split(\".\")[0] + \"_oneSNP.vcf\" # e.g., CG_l1_25_c86_H_biall_maf_oneSNP.vcf\n",
    "one_str_file = one_VCF[:-3] + \"str\" # e.g., CG_l1_25_c86_H_biall_maf_oneSNP.str\n",
    "str_inames_file = one_VCF[:-4] + \"_inames.str\" # e.g., CG_l1_25_c86_H_biall_maf_oneSNP_inames.str\n",
    "str_inames_fpop_file = one_VCF[:-4] + \"_inames_fpop.str\" # e.g., CG_l1_25_c86_H_biall_maf_oneSNP_inames_fpop.str\n",
    "\n",
    "# change wd\n",
    "os.chdir(dir_outfiles)\n",
    "\n",
    "# make directory for new output files, if doesn't already exist\n",
    "if not os.path.exists(\"new_outfiles\"):\n",
    "    os.makedirs(\"new_outfiles\")\n",
    "    \n",
    "# make VCFtools call string\n",
    "vcf_call = \"vcftools --vcf \" + ipy_VCF + \" \"\n",
    "if filter_biall == \"TRUE\":\n",
    "    vcf_call += \"--min-alleles 2 --max-alleles 2 \"\n",
    "if filter_MAF == \"TRUE\":\n",
    "    vcf_call += \"--maf .05 \"\n",
    "vcf_call += \"--out new_outfiles/\" + filt_VCF_out + \" \"\n",
    "vcf_call += \"--recode-INFO-all --recode\"\n",
    "\n",
    "# call VCFtools\n",
    "sp.call([vcf_call],shell=True)\n",
    "\n",
    "# change wd to new outfiles\n",
    "os.chdir(\"new_outfiles\")\n",
    "\n",
    "# make oneSNP call string\n",
    "one_call = \"python ../../oneSNP.py \" + filt_VCF_out_recode + \" \" + one_VCF\n",
    "\n",
    "# call oneSNP script to filter one SNP per RAD tag\n",
    "sp.call([one_call], shell=True)\n",
    "\n",
    "# double-check number of loci in VCF file\n",
    "vcf_file = open(one_VCF,\"r\") # change!\n",
    "vcf_file_lines = vcf_file.readlines()\n",
    "vcf_file.close()\n",
    "        \n",
    "locus_names = {}\n",
    "snp_name_count = 1\n",
    "\n",
    "for line in vcf_file_lines:\n",
    "    if line[0] != \"#\":\n",
    "        linelist = line.strip().split()\n",
    "        locus = linelist[0]\n",
    "        pos = linelist[1]\n",
    "        vcf_locus_name = locus + \"_\" + pos\n",
    "        str_locus_name = \"SNP_\" + str(snp_name_count)\n",
    "        snp_name_count += 1\n",
    "        locus_names[str_locus_name] = vcf_locus_name  \n",
    "        \n",
    "# get sample list and snp name list for rewriting header to structure file       \n",
    "sample_list = []\n",
    "snp_name_list = []\n",
    "rdd = {} \n",
    "\n",
    "for line in vcf_file_lines:\n",
    "    if line[0:1] == \"#C\":\n",
    "        linelist = line.strip().split()\n",
    "        sample_list += linelist[9:]\n",
    "    elif line[0] != \"#\": # ignore header lines\n",
    "        genblocks = line.strip().split()\n",
    "        snp_name = genblocks[0] + \"_\" + genblocks[1] # name format e.g., locus_47_1\n",
    "        snp_name_list.append(snp_name)\n",
    "        for genblock in genblocks[9:]: # start on genotype lines\n",
    "            genblocklist = genblock.split(\":\")\n",
    "            rd = int(genblocklist[1])\n",
    "            if snp_name not in rdd:\n",
    "                rdd[snp_name] = [rd]\n",
    "            elif snp_name in rdd:\n",
    "                rdd[snp_name] += [rd]\n",
    "            else:\n",
    "                print \"something funky going on\"\n",
    "                \n",
    "                \n",
    "# PGD Spider transform VCF to STR\n",
    "\n",
    "# change to PGD Spider directory\n",
    "os.chdir(\"/mnt/hgfs/SHARED_FOLDER/Software/PGDSpider_2.1.1.0\")\n",
    "\n",
    "# make PGD call string\n",
    "PGD_call = \"java -Xmx1024m -Xms512m -jar PGDSpider2-cli.jar \"\n",
    "PGD_call += \"-inputfile \" + dir_outfiles + \"/new_outfiles/\" + one_VCF\n",
    "PGD_call += \" -outputfile \" + dir_outfiles + \"/new_outfiles/\" + one_str_file\n",
    "PGD_call += \" -inputformat VCF -outputformat STRUCTURE \"\n",
    "PGD_call += \"-spid vcf_to_str.spid\"\n",
    "\n",
    "# call PGD to make STR file\n",
    "sp.call([PGD_call], shell = True)\n",
    "\n",
    "# change wd back to new outfiles\n",
    "os.chdir(dir_outfiles + \"/new_outfiles\")\n",
    "\n",
    "# rename structure file header\n",
    "old_str = open(one_str_file, \"r\")\n",
    "old_str_lines = old_str.readlines()\n",
    "old_str.close()\n",
    "\n",
    "new_str_header = \"\\t\\t\" # will need to change this line when I get a hold of better code for one SNP/tag\n",
    "old_str_header_list = old_str_lines[0].strip().split()\n",
    "for snpname in snp_name_list:\n",
    "    new_str_header += snpname + \" \"\n",
    "new_str_header = new_str_header[:-1]\n",
    "new_str = open(str_inames_file, \"w\")\n",
    "new_str.write(new_str_header + \"\\n\")\n",
    "for line in old_str_lines[1:]:\n",
    "    new_str.write(line)\n",
    "new_str.close()\n",
    "\n",
    "# add fake pop for analyzing populations together in hierfstat\n",
    "str_onesnp_file = open(str_inames_file, \"r\")\n",
    "str_onesnp_file_lines = str_onesnp_file.readlines()\n",
    "str_onesnp_file.close()\n",
    "\n",
    "fpop_lines = []\n",
    "\n",
    "for line in str_onesnp_file_lines[-6:]:\n",
    "    linelist = line.strip().split()\n",
    "    newline = \"FALSE\" + linelist[0] + \"\\t\" + \"2\" + \"\\t\"\n",
    "    for genotype in linelist[2:]:\n",
    "        newline += genotype + \" \"\n",
    "    newline = newline[:-1] + \"\\n\"\n",
    "    fpop_lines.append(newline)\n",
    "\n",
    "str_file_fpop =open(str_inames_fpop_file, \"w\")\n",
    "\n",
    "for line in str_onesnp_file_lines:\n",
    "    str_file_fpop.write(line)\n",
    "for line in fpop_lines:\n",
    "    str_file_fpop.write(line)\n",
    "str_file_fpop.close()\n",
    "\n",
    "numloci = len(str_onesnp_file_lines[0].strip().split())\n",
    "numinds = int(float(len(fpop_lines)+len(str_onesnp_file_lines)-1))/float(2)\n",
    "\n",
    "print \"Number of individuals =\" + str(numinds - 3)\n",
    "print \"Number of individuals WITH FAKE POP = \" + str(numinds) \n",
    "print \"Number of loci = \" + str(numloci)\n",
    "\n",
    "# make R call string\n",
    "string_callR = \"Rscript ../../../../pop_gen_stats.R \"\n",
    "string_callR += dir_outfiles + \"new_outfiles \"\n",
    "string_callR += str_inames_fpop_file + \" \"\n",
    "string_callR += str(numinds) + \" \" + str(numloci) + \" \" + date\n",
    "\n",
    "# run R script using call string\n",
    "sp.call([string_callR],shell=True)\n",
    "\n",
    "# names of pop gen stats files + name for plot title\n",
    "assembly_name_plots = assembly_name\n",
    "str_filename = str_inames_fpop_file\n",
    "fis_filename = \"Fis_\" + str_inames_fpop_file[:-4] + \"_\" + date + \".txt\"\n",
    "ho_filename = \"Ho_\" + str_inames_fpop_file[:-4] + \"_\" + date + \".txt\"\n",
    "hs_filename = \"Hs_\" + str_inames_fpop_file[:-4] + \"_\" + date + \".txt\"\n",
    "\n",
    "# make read depth plot\n",
    "vcf_file_fplot = open(one_VCF,\"r\") # change!\n",
    "vcf_file_lines = vcf_file_fplot.readlines()\n",
    "vcf_file.close()\n",
    "\n",
    "locus_names = {}\n",
    "snp_name_count = 1\n",
    "\n",
    "for line in vcf_file_lines:\n",
    "    if line[0] != \"#\":\n",
    "        linelist = line.strip().split()\n",
    "        locus = linelist[0]\n",
    "        pos = linelist[1]\n",
    "        vcf_locus_name = locus + \"_\" + pos\n",
    "        str_locus_name = \"SNP_\" + str(snp_name_count)\n",
    "        snp_name_count += 1\n",
    "        locus_names[str_locus_name] = vcf_locus_name  \n",
    "\n",
    "sample_list = []\n",
    "snp_name_list = []\n",
    "rdd = {} \n",
    "\n",
    "for line in vcf_file_lines:\n",
    "    if line[0:1] == \"#C\":\n",
    "        linelist = line.strip().split()\n",
    "        sample_list += linelist[9:]\n",
    "    elif line[0] != \"#\": # ignore header lines\n",
    "        genblocks = line.strip().split()\n",
    "        snp_name = genblocks[0] + \"_\" + genblocks[1] # name formate.g., locus_47_1\n",
    "        snp_name_list.append(snp_name)\n",
    "        for genblock in genblocks[9:]: # start on genotype lines\n",
    "            genblocklist = genblock.split(\":\")\n",
    "            rd = int(genblocklist[1])\n",
    "            if snp_name not in rdd:\n",
    "                rdd[snp_name] = [rd]\n",
    "            elif snp_name in rdd:\n",
    "                rdd[snp_name] += [rd]\n",
    "            else:\n",
    "                print \"something funky going on\"\n",
    "\n",
    "locus_rd_avs = []\n",
    "for snp in snp_name_list:\n",
    "    locus_rd_avs.append(np.mean(rdd[snp]))\n",
    "    \n",
    "plt.hist(locus_rd_avs, bins = np.arange(0,max(locus_rd_avs)+1,4)-2)\n",
    "\n",
    "title = one_VCF\n",
    "if len(plot_subtitle) > 1:\n",
    "    title += \"\\n\" + plot_subtitle\n",
    "plt.title(title)\n",
    "\n",
    "plt.xlabel(\"Read depth\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "\n",
    "plt.savefig('rd_' + filt_VCF_out + '.png')\n",
    "plt.close()\n",
    "\n",
    "# make Fis plot\n",
    "\n",
    "fis_file = open(fis_filename,\"r\")\n",
    "fis_file_lines = fis_file.readlines()\n",
    "fis_file.close()\n",
    "\n",
    "fis_list = []\n",
    "fis_pls_NAs = []\n",
    "\n",
    "for line in fis_file_lines[1:]:\n",
    "    fis = line.strip().split()[1]\n",
    "    fis_pls_NAs.append(fis)\n",
    "    if fis != \"NA\":\n",
    "        fis_list.append(float(fis))\n",
    "    \n",
    "plt.hist(fis_list, bins = np.arange(-1.2,1.2,.05)-.025)\n",
    "\n",
    "title = \"Fis in \" + filt_VCF_out\n",
    "if len(plot_subtitle) > 1:\n",
    "    title += \"\\n\" + plot_subtitle\n",
    "plt.title(title)\n",
    "\n",
    "plt.xlabel(\"Fis\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "\n",
    "plt.savefig(\"Fis_\" + filt_VCF_out + '.png')\n",
    "plt.close()\n",
    "\n",
    "# make Ho plot\n",
    "\n",
    "ho_file = open(ho_filename,\"r\")\n",
    "ho_file_lines = ho_file.readlines()\n",
    "ho_file.close()\n",
    "\n",
    "ho_list = []\n",
    "ho_and_NAs = []\n",
    "\n",
    "for line in ho_file_lines[1:]:\n",
    "    ho = line.strip().split()[1]\n",
    "    ho_and_NAs.append(ho)\n",
    "    if ho != \"NA\":\n",
    "        ho_list.append(float(ho))\n",
    "    \n",
    "plt.hist(ho_list, bins = np.arange(0,1.2,.05)-.025)\n",
    "\n",
    "\n",
    "title = \"Ho in \" + filt_VCF_out\n",
    "if len(plot_subtitle) > 1:\n",
    "    title += \"\\n\" + plot_subtitle\n",
    "plt.title(title)\n",
    "\n",
    "plt.xlabel(\"Ho\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.savefig(\"Ho in \" + filt_VCF_out + '.png')\n",
    "plt.close()\n",
    "\n",
    "# make Hs plot\n",
    "\n",
    "hs_file = open(hs_filename,\"r\")\n",
    "hs_file_lines = hs_file.readlines()\n",
    "hs_file.close()\n",
    "\n",
    "hs_list = []\n",
    "hs_and_NAs = []\n",
    "\n",
    "for line in hs_file_lines[1:]:\n",
    "    hs = line.strip().split()[1]\n",
    "    hs_and_NAs.append(hs)\n",
    "    if hs != \"NA\":\n",
    "        hs_list.append(float(hs))\n",
    "    \n",
    "plt.hist(hs_list, bins = np.arange(0,1.2,.025)-.0125)\n",
    "\n",
    "title = \"Hs in \" + filt_VCF_out\n",
    "if len(plot_subtitle) > 1:\n",
    "    title += \"\\n\" + plot_subtitle\n",
    "plt.title(title)\n",
    "\n",
    "plt.xlabel(\"Hs\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.savefig(\"Hs_\" + filt_VCF_out + '.png')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Check file names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CG_L1_L2_refgen_001.vcf\n",
      "CG_L1_L2_refgen_001_biall_NOmaf\n",
      "CG_L1_L2_refgen_001_biall_NOmaf.recode.vcf\n",
      "CG_L1_L2_refgen_001_biall_NOmaf_oneSNP.vcf\n",
      "CG_L1_L2_refgen_001_biall_NOmaf_oneSNP.str\n",
      "CG_L1_L2_refgen_001_biall_NOmaf_oneSNP_inames.str\n",
      "CG_L1_L2_refgen_001_biall_NOmaf_oneSNP_inames_fpop.str\n"
     ]
    }
   ],
   "source": [
    "print ipy_VCF\n",
    "print filt_VCF_out\n",
    "print filt_VCF_out_recode\n",
    "print one_VCF\n",
    "print one_str_file\n",
    "print str_inames_file\n",
    "print str_inames_fpop_file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
